File: apps/ios/Wordocious/Sources/InviteService.swift
import Foundation
import WordociousCore

/// Private-match invites — ports apps/web/lib/invite-service.ts. The socket
/// server pairs two clients purely on the `invite_code` string they pass to
/// join_queue; the `match_invites` row carries the game_mode so the joiner
/// knows which mode to launch (and powers cross-client web↔native invites).
enum InviteService {
    private static let alphabet = Array("ABCDEFGHJKLMNPQRSTUVWXYZ23456789") // no 0/O/1/I/l

    static func generateCode(length: Int = 8) -> String {
        String((0..<length).map { _ in alphabet[Int.random(in: 0..<alphabet.count)] })
    }

    private struct InviteInsert: Encodable {
        let inviter_id: String; let invite_code: String; let game_mode: String
    }
    private struct InviteInsertTargeted: Encodable {
        let inviter_id: String; let invitee_id: String?; let invite_code: String; let game_mode: String
    }
    private struct InviteRow: Decodable { let game_mode: String }
    private struct ProfileIdRow: Decodable { let id: String }
    private struct AcceptUpdate: Encodable {
        let status: String; let accepted_at: String; let match_id: String?
    }

    /// Result of the home Invite modal — a shareable code or a user-facing error.
    struct InviteResult { let code: String?; let error: String? }

    /// Create an invite, optionally targeted at a username (ports the web
    /// createInvite). A username is resolved to a profile id stored in
    /// `invitee_id` so that user sees it in their pending-invites badge; an
    /// empty username makes a public-link invite anyone can redeem.
    static func createInvite(gameMode: GameMode, inviteeUsername: String?) async -> InviteResult {
        let client = AuthService.shared.client
        guard let uid = (try? await client.auth.session.user.id.uuidString)?.lowercased() else {
            return InviteResult(code: nil, error: "You're not signed in")
        }
        var inviteeId: String?
        if let uname = inviteeUsername, !uname.isEmpty {
            let row: ProfileIdRow? = try? await client.from("profiles")
                .select("id").ilike("username", pattern: uname).limit(1).single().execute().value
            guard let row else { return InviteResult(code: nil, error: "User not found") }
            if row.id.lowercased() == uid { return InviteResult(code: nil, error: "You can't invite yourself") }
            inviteeId = row.id
        }
        for _ in 0..<3 {
            let code = generateCode()
            do {
                try await client.from("match_invites")
                    .insert(InviteInsertTargeted(inviter_id: uid, invitee_id: inviteeId, invite_code: code, game_mode: gameMode.rawValue))
                    .execute()
                return InviteResult(code: code, error: nil)
            } catch { continue }
        }
        return InviteResult(code: nil, error: "Could not generate a unique code")
    }

    /// Create an invite for a mode; returns the shareable code (nil on failure).
    /// Tries a few times on the unlikely unique-code collision.
    static func createInvite(gameMode: GameMode) async -> String? {
        let client = AuthService.shared.client
        guard let userId = try? await client.auth.session.user.id.uuidString else { return nil }
        for _ in 0..<3 {
            let code = generateCode()
            do {
                try await client.from("match_invites")
                    .insert(InviteInsert(inviter_id: userId, invite_code: code, game_mode: gameMode.rawValue))
                    .execute()
                return code
            } catch { continue }
        }
        return nil
    }

    /// A pending invite targeted at the current user (powers the home banner).
    struct PendingInvite: Decodable, Identifiable {
        let id: String
        let inviter_id: String
        let invite_code: String
        let game_mode: String
    }

    /// Invites sent TO this user that are still pending + unexpired — ports
    /// fetchPendingInvitesForUser.
    static func fetchPending(userId: String) async -> [PendingInvite] {
        let now = ISO8601DateFormatter().string(from: Date())
        return (try? await AuthService.shared.client.from("match_invites")
            .select("id,inviter_id,invite_code,game_mode")
            .eq("invitee_id", value: userId)
            .eq("status", value: "pending")
            .gt("expires_at", value: now)
            .order("created_at", ascending: false)
            .execute().value) ?? []
    }

    /// The inviter's display name for the banner.
    static func inviterUsername(_ inviterId: String) async -> String? {
        struct Row: Decodable { let username: String? }
        let row: Row? = try? await AuthService.shared.client.from("profiles")
            .select("username").eq("id", value: inviterId).limit(1).single().execute().value
        return row?.username
    }

    /// Decline an invite (dismiss from the banner).
    static func decline(inviteId: String) async {
        struct DeclineUpdate: Encodable { let status: String }
        try? await AuthService.shared.client.from("match_invites")
            .update(DeclineUpdate(status: "declined")).eq("id", value: inviteId).execute()
    }

    /// Look up the mode for an invite code (so the joiner launches the right mode).
    static func lookupMode(code: String) async -> GameMode? {
        let client = AuthService.shared.client
        let row: InviteRow? = try? await client.from("match_invites")
            .select("game_mode")
            .eq("invite_code", value: code)
            .limit(1).single()
            .execute().value
        return row.flatMap { GameMode(rawValue: $0.game_mode) }
    }

    /// Flip the pending invite to accepted once the server pairs both sides.
    static func markAccepted(code: String, matchId: String?) async {
        let client = AuthService.shared.client
        let iso = ISO8601DateFormatter().string(from: Date())
        try? await client.from("match_invites")
            .update(AcceptUpdate(status: "accepted", accepted_at: iso, match_id: matchId))
            .eq("invite_code", value: code)
            .eq("status", value: "pending")
            .execute()
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/InviteService.kt
package com.wordocious.app.data

import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * VS match invites — Android port of web lib/invite-service.ts (the parts the
 * pending-invites banner needs): list my pending un-expired invites, resolve
 * the inviter's username, decline.
 */
object InviteService {
    private val client get() = SupabaseConfig.client

    @Serializable
    data class MatchInvite(
        val id: String,
        @SerialName("inviter_id") val inviterId: String,
        @SerialName("invitee_id") val inviteeId: String? = null,
        @SerialName("invite_code") val inviteCode: String,
        @SerialName("game_mode") val gameMode: String,
        val status: String,
        @SerialName("expires_at") val expiresAt: String,
        @SerialName("created_at") val createdAt: String,
    )

    suspend fun fetchPendingInvitesForUser(userId: String): List<MatchInvite> = runCatching {
        client.postgrest["match_invites"]
            .select {
                filter {
                    eq("invitee_id", userId)
                    eq("status", "pending")
                    gt("expires_at", java.time.Instant.now().toString())
                }
                order("created_at", Order.DESCENDING)
            }
            .decodeList<MatchInvite>()
    }.getOrElse { emptyList() }

    @Serializable
    private data class NameIdRow(val id: String, val username: String? = null)

    /**
     * Batched inviter-username lookup: ONE profiles query for all inviter ids
     * (match_invites.inviter_id references auth.users, not profiles, so a
     * PostgREST embed isn't possible — this mirrors web/iOS). Returns id → username.
     */
    suspend fun lookupInviterUsernames(inviterIds: List<String>): Map<String, String> {
        val ids = inviterIds.distinct()
        if (ids.isEmpty()) return emptyMap()
        return runCatching {
            client.postgrest["profiles"]
                .select(Columns.raw("id,username")) { filter { isIn("id", ids) } }
                .decodeList<NameIdRow>()
                .mapNotNull { row -> row.username?.let { row.id to it } }
                .toMap()
        }.getOrElse { emptyMap() }
    }

    suspend fun markInviteDeclined(inviteId: String) {
        runCatching {
            client.postgrest["match_invites"].update({ set("status", "declined") }) {
                filter { eq("id", inviteId) }
            }
        }
    }

    @Serializable
    private data class GameModeRow(@SerialName("game_mode") val gameMode: String)

    /**
     * Which mode a shared invite code is for (joiner side) — web/iOS lookupMode.
     * Returns the game_mode string, or null if the code is unknown/expired.
     */
    suspend fun lookupMode(code: String): String? = runCatching {
        client.postgrest["match_invites"]
            .select(Columns.raw("game_mode")) {
                filter { eq("invite_code", code); eq("status", "pending") }
                limit(1)
            }
            .decodeSingleOrNull<GameModeRow>()?.gameMode
    }.getOrNull()

    /** Flip an invite to accepted once the server has paired both players. */
    suspend fun markInviteAccepted(code: String, matchId: String?) {
        runCatching {
            client.postgrest["match_invites"].update({
                set("status", "accepted")
                if (matchId != null) set("match_id", matchId)
            }) {
                filter { eq("invite_code", code); eq("status", "pending") }
            }
        }
    }

    // MARK: Outgoing invites (home Invite modal — web lib/invite-service.ts createInvite)

    /** Result of the home Invite modal — a shareable code or a user-facing error. */
    data class InviteResult(val code: String?, val error: String?)

    // No 0/O/1/I/l — easier to type (matches web CODE_ALPHABET + iOS).
    private const val CODE_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
    private fun generateCode(length: Int = 8): String =
        (0 until length).map { CODE_ALPHABET.random() }.joinToString("")

    @Serializable
    private data class IdRow(val id: String)

    @Serializable
    private data class InviteInsert(
        @SerialName("inviter_id") val inviterId: String,
        @SerialName("invitee_id") val inviteeId: String? = null,
        @SerialName("invite_code") val inviteCode: String,
        @SerialName("game_mode") val gameMode: String,
    )

    /**
     * Create a VS invite for a mode. If [inviteeUsername] is non-empty the invite
     * is targeted at that user (pending-invites badge); otherwise it's a public
     * link anyone can redeem. Returns the shareable code, or a user-facing error.
     * Retries a few times on the vanishingly unlikely unique-code collision.
     */
    suspend fun createInvite(gameMode: String, inviteeUsername: String?): InviteResult {
        val uid = AuthService.userId?.lowercase()
            ?: return InviteResult(null, "You're not signed in")

        var inviteeId: String? = null
        val uname = inviteeUsername?.trim()
        if (!uname.isNullOrEmpty()) {
            val row = runCatching {
                client.postgrest["profiles"]
                    .select(Columns.raw("id")) { filter { ilike("username", uname) }; limit(1) }
                    .decodeSingleOrNull<IdRow>()
            }.getOrNull() ?: return InviteResult(null, "User not found")
            if (row.id.lowercase() == uid) return InviteResult(null, "You can't invite yourself")
            inviteeId = row.id
        }

        repeat(3) {
            val code = generateCode()
            val ok = runCatching {
                client.postgrest["match_invites"]
                    .insert(InviteInsert(inviterId = uid, inviteeId = inviteeId, inviteCode = code, gameMode = gameMode))
            }.isSuccess
            if (ok) return InviteResult(code, null)
        }
        return InviteResult(null, "Could not generate a unique code")
    }
}
